### ISSUE TITLE
Travis

### ISSUE THREAD

**Issue by johnhamelink** [erts-io/erts_web#3]
Whilst working away today, I noticed that travis has precompiled binary repository that it downloads as part of the build service.

Example:
- https://s3.amazonaws.com/travis-otp-releases/binaries/ubuntu/12.04/x86_64/erlang-18.2.1-nonroot.tar.bz2

I wonder if it'd be possible for Travis to produce a list of releases for us - that'd give us all we need to have builds - they've already produced everything we need!

**Comment by johnhamelink** [comment]
See here for reference: https://twitter.com/john_hamelink/status/704989912026779648

**Comment by andrewvy** [comment]
Looks like relx needs the path to the whole erlang folder, and we can't just bundle the `erts` folder like in [erlang/otp](https://github.com/erlang/otp)

But I'm not sure, the documentation doesn't mention clearly: https://github.com/erlware/relx/wiki/configuration#alternate-erts

Here is the code for how it includes erts when you explicitly specify it: https://github.com/erlware/relx/blob/e29f0ec6820abbf692fa878f8fb284911d5a3813/src/rlx_prv_assembler.erl#L467

It looks like specifying the path in `include_erts` adds the `erts` folder path sadly

```
ErtsDir = filename:join([Prefix, "erts-" ++ ErtsVersion]),
```

**Comment by johnhamelink** [comment]
@tsloughter - any advice on this?

**Comment by bitwalker** [comment]
@andrewvy You need the lib directory from Erlang, i.e. `/usr/lib/erlang/lib`, for `{include_erts, "path"}.`. Any cross-compiled package will need them as well. The reason for this is that some standard library Erlang modules depend on natively-implemented functions, such as `crypto`, so they have to be compiled for a specific platform/architecture, and some (all?) are also dynamically linked against libraries, e.g. OpenSSL for `crypto` (and others I believe). So a cross-compiled package is specific to:

- The host CPU architecture
- The host operating system
- The versions of certain host libraries (OpenSSL is the only one I know of off hand, perhaps there are others, a review of the erlang/otp repository would show for sure)

I think the way we'd want to handle this is with a fairly restricted matrix at first, something like:

| CPU | OS | OpenSSL |
|----------|-----|---------------|
| x86_64 | Ubuntu Stable | 1.0.2.g |
| x86_64 | Debian Stable | 1.0.2g |
| x86_64 | Fedora Stable | 1.0.2g |
| x86_64 | Raspbian Stable | 1.0.2g |

Maybe with CentOS/RedHat too, anyway, start with something restricted at first, then expand it as it becomes easier to do so.

**Comment by tsloughter** [comment]
Not exactly sure what is trying to be accomplished but to include `crypto` you need to also specify `{system_libs, "path/to/erlang/libs"}` :)

**Comment by bitwalker** [comment]
@tsloughter Ah yep, missed that, thanks :)

**Comment by andrewvy** [comment]
@bitwalker 

Thanks for the detailed reply! :)

So, it does look like we will have to provide precompiled erlang binaries then. `OpenSSL` seems very interesting, as that's one dependency that will have to be monitored and updated when needed. From a quick research, I think that's the only other dependency that's really needed as its used for crypto and ssl. (Unless you wanted WxWidgets.)

I wish TravisCI could give us some details. We can definitely figure out their naming structure, but I'd rather just build the Jenkins cluster for compilation. Though, we could have cloned their binaries and re-host them, postponing the build system infrastructure for a bit.

**Comment by bitwalker** [comment]
@andrewvy Yeah, what's required is still a subset of the complete Erlang installation, but it's naturally a significant subset, and as pointed out by Tristan, you also have to bundle the system libs. That said, it's still a pretty compact package, and by doing this, it becomes very easy to get your project deployed to another system.

Oh, and yeah I wouldn't bother with WxWidgets initially, people that need cross-compiled releases with that can tackle that one manually for the time being.

If you already have access to a Jenkins cluster, that's probably easier to get going with than trying to guess Travis's prebuilt URLs, but if they open up an API or a metadata file at some point, that will be the nicer route I think.
